File: MatlabCodes/Chapter 09/Fig09_14.m
function ImportanceSampling
clc;close all
format short

% Exact solution P(X>5)

a        = 3;
exactSol = 1-normcdf(a,0,1);

% Specification of a grid for the number of samples

gridN = 1000:100:100000; 

% Creation of empty vectors to store the output

Solution  =  zeros(length(gridN),1);
SolutionCase1  =  zeros(length(gridN),1);
SolutionCase2  =  zeros(length(gridN),1);
SolutionCase3  =  zeros(length(gridN),1);
SolutionCase4  =  zeros(length(gridN),1);

% Function holds for Radon-Nikodym derivative

L1 = @(x)normpdf(x,0,1)./unifpdf(x,0,1);
L2 = @(x)normpdf(x,0,1)./unifpdf(x,0,4);
L3 = @(x)normpdf(x,0,1)./normpdf(x,0,0.5);
L4 = @(x)normpdf(x,0,1)./normpdf(x,0,3);

% Running index;

i=1;

for N=gridN

    % Standard approach

    X = random('normal',0,1,[N,1]);
    RV = double(X>a);
    Pr=mean(RV);
    Solution(i)=Pr;
    
    % Case 1, uniform(0,1)

    Y = random('uniform',0,1,[N,1]);
    RV = double(Y>a).*L1(Y);
    Pr=mean(RV);
    SolutionCase1(i)=Pr;
    
    % Case 2, uniform(0,4)

    Y = random('uniform',0,4,[N,1]);
    RV = double(Y>a).*L2(Y);
    Pr=mean(RV);
    SolutionCase2(i)=Pr;
    
    % Case 3, normal(0,0.5)

    Y = random('normal',0,0.5,[N,1]);
    RV = double(Y>a).*L3(Y);
    Pr=mean(RV);
    SolutionCase3(i)=Pr;
    
    % Case 4, normal(0,3)

    Y = random('normal',0,3,[N,1]);
    RV = double(Y>a).*L4(Y);
    Pr=mean(RV);
    SolutionCase4(i)=Pr;
    
    i=i+1;
end
%plot(gridN,Variance,'-r','linewidth',2)
plot(gridN,Solution,'.')
hold on
plot(gridN,SolutionCase1,'.k')
plot(gridN,SolutionCase2,'.g')
plot(gridN,SolutionCase3,'.m')
plot(gridN,SolutionCase4,'.y')
plot(gridN,exactSol*ones(length(gridN),1),'-r','linewidth',2)
grid on
xlabel('Number of samples')
ylabel('Estimated value')
legend('Standard Monte Carlo','U([0,1])','U([0,4])','N(0,0.25)', 'N(0,9)','exact')
    

File: PythonCodes/Chapter 09/Fig09_14.py
#%%
"""
Importance Sampling exercise
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

# Global variable 

a = 3

# Benchmark

def ExactValue():
    
    val = 1.0- st.norm.cdf(a,0,1)
    print("Reference value is equal to %.2f" %val)
    return val   
        
# Monte Carlo simulation of the problem

def MonteCarloSimulation():    

    # Specification of the grid for the number of samples

    gridN = range(1000,100000,100)
    
    solution = np.zeros([len(gridN)])
    solutionCase_1 = np.zeros([len(gridN)])
    solutionCase_2 = np.zeros([len(gridN)])
    solutionCase_3 = np.zeros([len(gridN)])
    solutionCase_4 = np.zeros([len(gridN)])
    
    # Lambda expressions for Radon-Nikodym derivative

    L1 = lambda x: st.norm.pdf(x,0.0,1.0)/st.uniform.pdf(x,0.0,1.0)
    L2 = lambda x: st.norm.pdf(x,0.0,1.0)/st.uniform.pdf(x,0.0,4.0)
    L3 = lambda x: st.norm.pdf(x,0.0,1.0)/st.norm.pdf(x,0.0,0.5)
    L4 = lambda x: st.norm.pdf(x,0.0,1.0)/st.norm.pdf(x,0.0,3.0)
    
    #Generate results 

    for idx, N in enumerate(gridN):

        # Direct Monte Carlo simulation

        X = np.random.normal(0.0,1.0,N)
        RV = X>a
        solution[idx] = np.mean(RV)
        
        # Case 1, uniform (0,1)

        Y_temp = np.random.uniform(0.0,1.0,N)
        RV = (Y_temp>a) * 1.0 * L1(Y_temp)
        solutionCase_1[idx] = np.mean(RV)
        
        # Case 2, uniform (0,4)

        Y_temp = np.random.uniform(0.0,4.0,N)
        RV = (Y_temp>a) * 1.0 * L2(Y_temp)
        solutionCase_2[idx] = np.mean(RV)
        
        # Case 3, normal (0,0.5)

        Y_temp = np.random.normal(0.0,0.5,N)
        RV = (Y_temp>a) * 1.0 * L3(Y_temp)
        solutionCase_3[idx] = np.mean(RV)
        
        # Case 4, normal (0,3)

        Y_temp = np.random.normal(0.0,3.0,N)
        RV = (Y_temp>a) * 1.0 * L4(Y_temp)
        solutionCase_4[idx] = np.mean(RV)
        
    plt.plot(gridN,ExactValue() * np.ones(len(gridN)))
    plt.plot(gridN,solution)
    plt.plot(gridN,solutionCase_1)
    plt.plot(gridN,solutionCase_2)
    plt.plot(gridN,solutionCase_3)
    plt.plot(gridN,solutionCase_4)
    plt.xlabel("Number of Samples")
    plt.ylabel("Estimated value")
    plt.grid()
    plt.legend(["exact","Monte Carlo","Case 1", "Case 2", "Case 3", "Case 4"])
